src/pointer/ptr.rs

```diff
@@ -0,0 +1,1010 @@
+    {
+        /// Constructs a `Ptr` from a [`NonNull`].
+        ///
+        /// # Safety
+        ///
+        /// The caller promises that:
+        ///
+        /// 0. `ptr` is derived from some valid Rust allocation, `A`.
+        /// 1. `ptr` has valid provenance for `A`.
+        /// 2. `ptr` addresses a byte range which is entirely contained in `A`.
+        /// 3. `ptr` addresses a byte range whose length fits in an `isize`.
+        /// 4. `ptr` addresses a byte range which does not wrap around the
+        ///    address space.
+        /// 5. `A` is guaranteed to live for at least `'a`.
+        /// 6. `ptr` conforms to the aliasing invariant of
+        ///   [`ALIASING_INVARIANT`](invariant::Aliasing).
+        /// 7. `ptr` conforms to the alignment invariant of
+        ///   [`ALIGNMENT_INVARIANT`](invariant::Alignment).
+        /// 8. `ptr` conforms to the validity invariant of
+        ///   [`VALIDITY_INVARIANT`](invariant::Validity).
+        pub(super) const unsafe fn new(
+            ptr: NonNull<T>,
+        ) -> Ptr<'a, T, ALIASING_INVARIANT, ALIGNMENT_INVARIANT, VALIDITY_INVARIANT> {
+            // SAFETY: The caller has promised to satisfy all safety invariants
+            // of `Ptr`.
+            Self { ptr, _lifetime: PhantomData }
+        }
+
+        /// Constructs a `Ptr` from another `Ptr`, possibly with different
+        /// parameterized safety invariants.
+        ///
+        /// # Safety
+        ///
+        /// The caller promises that:
+        /// - `ptr` conforms to the aliasing invariant of
+        ///   [`ALIASING_INVARIANT`](invariant::Aliasing).
+        /// - `ptr` conforms to the alignment invariant of
+        ///   [`ALIGNMENT_INVARIANT`](invariant::Alignment).
+        /// - `ptr` conforms to the validity invariant of
+        ///   [`VALIDITY_INVARIANT`](invariant::Validity).
+        pub(super) const unsafe fn from_ptr<
+            const INITIAL_ALIASING_INVARIANT: invariant::Aliasing,
+            const INITIAL_ALIGNMENT_INVARIANT: invariant::Alignment,
+            const INITIAL_VALIDITY_INVARIANT: invariant::Validity,
+        >(
+            ptr: Ptr<
+                'a,
+                T,
+                INITIAL_ALIASING_INVARIANT,
+                INITIAL_ALIGNMENT_INVARIANT,
+                INITIAL_VALIDITY_INVARIANT,
+            >,
+        ) -> Self {
+            // SAFETY: The caller has promised to satisfy all parameterized
+            // invariants of `Ptr`. `Ptr`'s other invariants are satisfied
+            // by-contract by the source `Ptr`.
+            unsafe { Self::new(ptr.as_non_null()) }
+        }
+
+        /// Converts this `Ptr<T>` to a [`NonNull<T>`].
+        ///
+        /// Note that this method does not consume `self`. Callers should be
+        /// aware that safe code may assume that the aliasing invariants of
+        /// `self` are not violated, and that the `NonNull` may be used in ways
+        /// that violate those assumptions.
```

Comment:
Did you mean that *unsafe* code may assume that the aliasing invariants aren't violated? Also, did you mean to warn that it's possible to use the returned `NonNull` in a way that violates those invariants (as opposed to warning that "safe code may assume... that the `NonNull` may be used in ways that violate those assumptions")?